src/control/pbin/app.go

```diff
@@ -0,0 +1,245 @@
+		WithLogLevel(logLevel)
+}
+
+// WithAllowedCallers adds a list of process names allowed to call this
+// application.
+func (a *App) WithAllowedCallers(callers ...string) *App {
+	a.allowedCallers = callers
+	return a
+}
+
+// WithInput adds a custom input source to the App.
+func (a *App) WithInput(reader io.ReadCloser) *App {
+	a.input = reader
+	return a
+}
+
+// WithOutput adds a custom output sink to the App.
+func (a *App) WithOutput(writer io.WriteCloser) *App {
+	a.output = writer
+	return a
+}
+
+// AddHandler adds a new handler to the App for a given method.
+// There is at most one handler per method.
+func (a *App) AddHandler(method string, handler RequestHandler) {
+	a.handlers[method] = handler
+}
+
+// logError is a convenience method that logs an error and returns the same error.
+func (a *App) logError(err error) error {
+	if a.log != nil && err != nil {
+		a.log.Error(err.Error())
+	}
+	return err
+}
+
+// Run executes the helper application process.
+func (a *App) Run() error {
+	parentName, err := a.process.ParentProcessName()
+	if err != nil {
+		return a.logError(err)
+	}
+
+	if !a.isCallerPermitted(parentName) {
+		return a.logError(errors.Errorf("%s (version %s) may only be invoked by: %s",
+			a.Name(), build.DaosVersion, strings.Join(a.allowedCallers, ", ")))
+	}
+
+	// set up the r/w pipe from the parent process
+	conn := NewStdioConn(a.Name(), parentName, a.input, a.output)
+
+	if err = a.checkPrivileges(); err != nil {
+		resp := NewResponseWithError(a.logError(err))
+		sendErr := a.writeResponse(resp, conn)
+		if sendErr != nil {
+			return a.logError(sendErr)
+		}
+
+		return err
+	}
+
+	req, err := a.readRequest(conn)
+	if err != nil {
+		return a.logError(err)
+	}
+
+	resp := a.handleRequest(req)
+
+	err = a.writeResponse(resp, conn)
+	if err != nil {
+		return a.logError(err)
+	}
+
+	if resp.Error != nil {
+		return resp.Error
+	}
+	return nil
+}
+
+func (a *App) isCallerPermitted(callerName string) bool {
+	if len(a.allowedCallers) == 0 {
+		return true
+	}
+	for _, name := range a.allowedCallers {
+		if callerName == name {
+			return true
+		}
+	}
+	return false
+}
+
+func (a *App) checkPrivileges() error {
+	if !a.process.IsPrivileged() {
+		return PrivilegedHelperNotPrivileged(a.Name())
+	}
+
+	// hack for stuff that doesn't use geteuid() (e.g. ipmctl)
+	if err := a.process.ElevatePrivileges(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (a *App) readRequest(rdr io.Reader) (*Request, error) {
+	buf, err := ReadMessage(rdr)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to read request")
+	}
+
+	var req Request
+	if err := json.Unmarshal(buf, &req); err != nil {
+		return nil, errors.Wrap(err, "failed to unmarshal request")
+	}
+
+	return &req, nil
+}
+
+func (a *App) handleRequest(req *Request) *Response {
+	reqHandler, ok := a.handlers[req.Method]
+	if !ok {
+		err := a.logError(errors.Errorf("unhandled method '%s'", req.Method))
```

Comment:
minor: The %q format directive will quote strings for you.

Reply:
Fixed